### PR TITLE
Stop leaking iproxy processes

### LINKED ISSUE #41854: macOS catalina is leaking processes
We're not exactly sure when this started happening, because the processes need to build up significantly before it causes test failures. Locally @gspencergoog, @christopherfujino, and I observed dozens of dart processes and hundreds of iproxy processes. killall with no exit code didn't work, but once we had manually killed them the tasks started passing again.

### PR DESCRIPTION
**Pull request by christopherfujino**
## Description

This actually fixes two issues:

1. #41854, that there are situations when we spawn many subprocesses of `iproxy` and never close them (even after the tool exits), which build up in the devicelab until the processes have to be manually killed.
2. #41911 That in many cases, we were port forwarding twice, once through the log scanning flow and once through the mDNS discovery flow.

This PR adds a new method to `Device` class, called `void killSubProcesses()`, which has an empty default implementation, but for an `IOSDevice` will kill `iproxy` and `idevicesyslog` processes, if they exist. I imagine this method would be useful for other types of devices (I currently have a stray `adb -L fork-server server` process running).

## Related Issues

Fixes #41854 & #41911

## Tests

Add unit tests for:

1. That `ForwardedPort` calling `.dispose()` will send term signal to the attached process
1. `.dispose()` method for `IOSDevice` will call `.dispose()` to any port forwarders or log readers
1. That `startApp()` kills first process before invoking second.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*
- [x] No, this is *not* a breaking change.

### CODE REVIEW

File: packages/flutter_tools/lib/src/ios/devices.dart [hunk 1]
```diff
@@ -634,6 +646,7 @@ class _IOSDevicePortForwarder extends DevicePortForwarder {
       // TODO(ianh): This is a flakey race condition, https://github.com/libimobiledevice/libimobiledevice/issues/674
       connected = !await process.stdout.isEmpty.timeout(_kiProxyPortForwardTimeout, onTimeout: () => false);
       if (!connected) {
+        process.kill();
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/ios/devices.dart]: Before this change, inside this while loop, we could spawn...enough `iproxy` instances here to exhaust the available file descriptors (as each tries to bind to a port).

File: packages/flutter_tools/lib/src/protocol_discovery.dart [hunk 2]
```diff
@@ -48,7 +48,10 @@ class ProtocolDiscovery {
   StreamSubscription<String> _deviceLogSubscription;
 
   /// The discovered service URI.
-  Future<Uri> get uri => _completer.future;
+  Future<Uri> get uri async {
+    final Uri rawUri = await _completer.future;
+    return await _forwardPort(rawUri);
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/protocol_discovery.dart]: Specifically only try to `_forwardPort` when asked for, because if mDNS discovery succeeds, we'll never use this.

File: packages/flutter_tools/lib/src/ios/devices.dart [hunk 3]
```diff
@@ -548,14 +560,14 @@ class _IOSDeviceLogReader extends DeviceLogReader {
 
   void _start() {
     iMobileDevice.startLogger(device.id).then<void>((Process process) {
-      _process = process;
-      _process.stdout.transform<String>(utf8.decoder).transform<String>(const LineSplitter()).listen(_newLineHandler());
-      _process.stderr.transform<String>(utf8.decoder).transform<String>(const LineSplitter()).listen(_newLineHandler());
-      _process.exitCode.whenComplete(() {
+      process.stdout.transform<String>(utf8.decoder).transform<String>(const LineSplitter()).listen(_newLineHandler());
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/ios/devices.dart]: Since these pids are created by the log reader/portfowarder implementations, they should be disposed by those implementations as well - adding `killSubProcesses` feels a bit leaky. I would suggest a slightly different arrangement:

1. Add a `dispose` method to the LogReader and PortForwarder interfaces.
2. Either cache the reader/port forwarder in the device, and likewise add a dispose to device, or otherwise hold onto these objects to dispose them.

File: packages/flutter_tools/lib/src/protocol_discovery.dart [hunk 4]
```diff
@@ -48,7 +48,10 @@ class ProtocolDiscovery {
   StreamSubscription<String> _deviceLogSubscription;
 
   /// The discovered service URI.
-  Future<Uri> get uri => _completer.future;
+  Future<Uri> get uri async {
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/protocol_discovery.dart]: Is this change necessary?

Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/protocol_discovery.dart]: Ahh, I missed that you moved `_forwardPort`. my bad!

File: packages/flutter_tools/lib/src/ios/devices.dart (hunk 3 above)
Comment (christopherfujino) [reply packages/flutter_tools/lib/src/ios/devices.dart]: done

File: packages/flutter_tools/lib/src/device.dart [hunk 5]
```diff
@@ -473,6 +474,8 @@ abstract class Device {
   static Future<void> printDevices(List<Device> devices) async {
     await descriptions(devices).forEach(printStatus);
   }
+
+  void dispose() {}
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/device.dart]: Nit: add a doc comment on all of the non-override dispose methods

File: packages/flutter_tools/lib/src/device.dart [hunk 6]
```diff
@@ -564,6 +567,18 @@ class ForwardedPort {
 
   @override
   String toString() => 'ForwardedPort HOST:$hostPort to DEVICE:$devicePort';
+
+  bool killProcess() {
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/device.dart]: Nit: doc comment here

**Comment by jonahwilliams** [comment]
Just a few nits. The most important part of this change will be the tests which ensure we don't regress here.

File: packages/flutter_tools/lib/src/device.dart (hunk 5 above)
Comment (christopherfujino) [reply packages/flutter_tools/lib/src/device.dart]: done

File: packages/flutter_tools/test/general.shard/ios/devices_test.dart [hunk 7]
```diff
@@ -75,6 +79,34 @@ void main() {
...
+        logReader1.idevicesyslogProcess = mockProcess1;
+        logReader2.idevicesyslogProcess = mockProcess2;
+        final IOSDevicePortForwarder portForwarder = IOSDevicePortForwarder(device);
+        final MockForwardedPort mockForwardedPort = MockForwardedPort();
+        portForwarder.addForwardedPorts(<ForwardedPort>[mockForwardedPort]);
+        device.setLogReader(applicationPackage1, logReader1);
+        device.setLogReader(applicationPackage2, logReader2);
+        device.portForwarder = portForwarder;
+        device.dispose();
+        verify(mockProcess1.kill());
```
Comment (jonahwilliams) [inline packages/flutter_tools/test/general.shard/ios/devices_test.dart]: Nit: there is a lot of code here, it would be easier to read if you separate the setup, the actions, and the expectations.

**Review by jonahwilliams — APPROVED** [review]
LGTM with nit

File: packages/flutter_tools/test/general.shard/ios/devices_test.dart (hunk 7 above)
Comment (christopherfujino) [reply packages/flutter_tools/test/general.shard/ios/devices_test.dart]: done
